<div style="
    margin:25px 0 30px 0;
    padding:0;
    border-radius:24px;
    overflow:hidden;
    background:#FFFFFF;
    border:1px solid #CFE5DC;
    box-shadow:0 10px 30px rgba(15,118,110,0.10);
    font-family:'Segoe UI',Arial,sans-serif;
    width:100%;
    box-sizing:border-box;
">

<div style="
    background:linear-gradient(135deg,#164E43 0%,#0F766E 55%,#138A78 100%);
    padding:34px 38px 30px 38px;
    position:relative;
">

<div style="
    display:flex;
    justify-content:space-between;
    align-items:center;
    margin-bottom:26px;
">

<div style="
    color:#DDF7EE;
    font-size:13px;
    font-weight:900;
    letter-spacing:3px;
">
BINX TECH · AI / ML INTERNSHIP
</div>

<div style="
    padding:8px 14px;
    border-radius:12px;
    background:rgba(255,255,255,0.12);
    border:1px solid rgba(255,255,255,0.25);
    color:#FFFFFF;
    font-size:12px;
    font-weight:900;
    letter-spacing:1px;
">
SPRINT 2
</div>

</div>

<div style="
    color:#FFFFFF;
    font-size:38px;
    font-weight:900;
    line-height:1.2;
    margin-bottom:12px;
">
Week 7 · Day 5
</div>

<div style="
    color:#DDF7EE;
    font-size:22px;
    font-weight:700;
    line-height:1.5;
">
Hands-On Lab
</div>

<div style="
    width:70px;
    height:4px;
    border-radius:10px;
    background:#B7E4D2;
    margin-top:22px;
">
</div>

</div>

<div style="
    padding:24px 38px 28px 38px;
    background:linear-gradient(135deg,#F8FCFA 0%,#FFFFFF 60%,#F8F5EE 100%);
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    letter-spacing:2px;
    margin-bottom:8px;
">
SPRINT 2 CLOSE-OUT
</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:10px;
">
From Experiments to Final Model
</div>

<p style="
    color:#607A73;
    font-size:15px;
    font-weight:600;
    line-height:1.8;
    margin:0;
    max-width:900px;
">
A practical close-out of Sprint 2, covering model improvement,
hyperparameter experiments, evaluation, comparison with the baseline,
and preparation of the final CardioML results.
</p>

<div style="
    display:flex;
    gap:10px;
    flex-wrap:wrap;
    margin-top:20px;
">

<div style="
    padding:8px 13px;
    border-radius:10px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    color:#0F766E;
    font-size:12px;
    font-weight:800;
">
MODEL IMPROVEMENT
</div>

<div style="
    padding:8px 13px;
    border-radius:10px;
    background:#F5F7F6;
    border:1px solid #D8E5DF;
    color:#42645B;
    font-size:12px;
    font-weight:800;
">
EXPERIMENTS
</div>

<div style="
    padding:8px 13px;
    border-radius:10px;
    background:#F5F7F6;
    border:1px solid #D8E5DF;
    color:#42645B;
    font-size:12px;
    font-weight:800;
">
EVALUATION
</div>

<div style="
    padding:8px 13px;
    border-radius:10px;
    background:#F5F7F6;
    border:1px solid #D8E5DF;
    color:#42645B;
    font-size:12px;
    font-weight:800;
">
FINAL REVIEW
</div>

</div>

</div>

</div>

<div style="
    margin:22px 0 30px 0;
    padding:28px 32px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    display:flex;
    justify-content:space-between;
    align-items:center;
    gap:15px;
    margin-bottom:14px;
    max-width:100%;
    box-sizing:border-box;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:2px;
    min-width:0;
">
STEP 1 · ARCHITECTURE DECISION
</div>

<div style="
    padding:7px 13px;
    border-radius:10px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    color:#0F766E;
    font-size:12px;
    font-weight:900;
    white-space:nowrap;
    box-sizing:border-box;
">
CARDIOML
</div>

</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:14px;
    max-width:100%;
    box-sizing:border-box;
">
Confirming the Core Architecture
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.85;
    font-weight:600;
    margin:0;
    max-width:100%;
    box-sizing:border-box;
">
CardioML is a structured tabular dataset containing patient-level
medical and lifestyle features such as age, gender, height, weight,
blood pressure, cholesterol, glucose, smoking, alcohol consumption,
and physical activity. The target variable, <b>cardio</b>, represents
a binary cardiovascular disease classification task.
</p>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.85;
    font-weight:600;
    margin:14px 0 0 0;
    max-width:100%;
    box-sizing:border-box;
">
Because the data is tabular rather than image-based, spatial or
sequential architectures such as CNNs, LSTMs, or GRUs are not the
natural choice for the core model. A <b>Dense Neural Network</b> is
appropriate because it can learn non-linear relationships between
the input features and the binary target.
</p>

<div style="
    margin-top:20px;
    padding:18px 20px;
    border-radius:14px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:1px;
    margin-bottom:8px;
">
ARCHITECTURE DECISION
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.7;
    font-weight:700;
    margin:0;
    max-width:100%;
    box-sizing:border-box;
">
✓ Data Type: Tabular Medical Data<br>
✓ Task: Binary Classification<br>
✓ Core Architecture: Dense Neural Network<br>
✓ Output Layer: Sigmoid Activation
</p>

</div>

</div>

<div style="
    margin:22px 0 24px 0;
    padding:26px 30px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:2px;
    margin-bottom:10px;
">
STEP 2 · MODEL TRAINING & TUNING
</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:12px;
">
Train and Tune the Improved Core Model
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.85;
    font-weight:600;
    margin:0;
">
The improved CardioML Dense Neural Network will be trained using
multiple hyperparameter configurations. Each experiment will use the
same data split and preprocessing pipeline so that the results can be
compared fairly.
</p>

<div style="
    margin-top:18px;
    padding:16px 18px;
    border-radius:14px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:1px;
    margin-bottom:7px;
">
EXPERIMENT LOGGING
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.7;
    font-weight:700;
    margin:0;
">
Each experiment records its learning rate, batch size, number of
epochs, best validation accuracy, validation loss, test accuracy,
and test loss.
</p>

</div>

</div>

In [1]:
experiments = [
    {
        "name": "Experiment 1",
        "learning_rate": 0.001,
        "batch_size": 64,
        "epochs": 3
    },
    {
        "name": "Experiment 2",
        "learning_rate": 0.0005,
        "batch_size": 64,
        "epochs": 5
    },
    {
        "name": "Experiment 3",
        "learning_rate": 0.0001,
        "batch_size": 32,
        "epochs": 5
    }
]

print("Hyperparameter Experiments:")

for exp in experiments:
    print(
        f"{exp['name']} | "
        f"Learning Rate: {exp['learning_rate']} | "
        f"Batch Size: {exp['batch_size']} | "
        f"Epochs: {exp['epochs']}"
    )

Hyperparameter Experiments:
Experiment 1 | Learning Rate: 0.001 | Batch Size: 64 | Epochs: 3
Experiment 2 | Learning Rate: 0.0005 | Batch Size: 64 | Epochs: 5
Experiment 3 | Learning Rate: 0.0001 | Batch Size: 32 | Epochs: 5


In [3]:
import pandas as pd
experiment_results = []

print("Experiment tracking table initialized.")

results_df = pd.DataFrame(
    experiment_results,
    columns=[
        "Experiment",
        "Learning Rate",
        "Batch Size",
        "Epochs",
        "Best Validation Accuracy",
        "Best Validation Loss",
        "Test Accuracy",
        "Test Loss"
    ]
)

results_df

Experiment tracking table initialized.


,Experiment,Learning Rate,Batch Size,Epochs,Best Validation Accuracy,Best Validation Loss,Test Accuracy,Test Loss


In [10]:
experiment_results = []

early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="val_loss",
    patience=8,
    restore_best_weights=True
)

reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(
    monitor="val_loss",
    factor=0.5,
    patience=3,
    min_lr=1e-6
)

experiment_model = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(64, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(32, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

experiment_model.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

experiment_history = experiment_model.fit(
    X_train_scaled,
    y_train,
    validation_data=(X_val_scaled, y_val),
    epochs=30,
    batch_size=64,
    callbacks=[early_stopping, reduce_lr],
    verbose=1
)

test_loss, test_accuracy = experiment_model.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

experiment_results.append({
    "Experiment": "Experiment 1",
    "Learning Rate": 0.001,
    "Batch Size": 64,
    "Max Epochs": 30,
    "Best Validation Accuracy": max(
        experiment_history.history["val_accuracy"]
    ),
    "Best Validation Loss": min(
        experiment_history.history["val_loss"]
    ),
    "Test Accuracy": test_accuracy,
    "Test Loss": test_loss
})

results_df = pd.DataFrame(experiment_results)

results_df

Epoch 1/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.6407 - loss: 0.6524 - val_accuracy: 0.6988 - val_loss: 0.5870 - learning_rate: 0.0010
Epoch 2/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7092 - loss: 0.5778 - val_accuracy: 0.7192 - val_loss: 0.5573 - learning_rate: 0.0010
Epoch 3/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7240 - loss: 0.5621 - val_accuracy: 0.7293 - val_loss: 0.5524 - learning_rate: 0.0010
Epoch 4/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7266 - loss: 0.5557 - val_accuracy: 0.7244 - val_loss: 0.5522 - learning_rate: 0.0010
Epoch 5/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7286 - loss: 0.5540 - val_accuracy: 0.7271 - val_loss: 0.5483 - learning_rate: 0.0010
Epoch 6/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7299 - loss: 0.5511 - val_accuracy: 0.7258 - val_loss: 0.5496 - learning_rate: 0.0010
Epoch 7/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7311 - loss: 0.5508 - 

,Experiment,Learning Rate,Batch Size,Max Epochs,Best Validation Accuracy,Best Validation Loss,Test Accuracy,Test Loss
0,Experiment 1,0.001,64,30,0.734381,0.543304,0.733619,0.547319


In [11]:
experiment_model_2 = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(64, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(32, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

experiment_model_2.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.0005),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

experiment_history_2 = experiment_model_2.fit(
    X_train_scaled,
    y_train,
    validation_data=(X_val_scaled, y_val),
    epochs=30,
    batch_size=64,
    callbacks=[
        tf.keras.callbacks.EarlyStopping(
            monitor="val_loss",
            patience=8,
            restore_best_weights=True
        ),
        tf.keras.callbacks.ReduceLROnPlateau(
            monitor="val_loss",
            factor=0.5,
            patience=3,
            min_lr=1e-6
        )
    ],
    verbose=1
)

test_loss_2, test_accuracy_2 = experiment_model_2.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

experiment_results.append({
    "Experiment": "Experiment 2",
    "Learning Rate": 0.0005,
    "Batch Size": 64,
    "Max Epochs": 30,
    "Best Validation Accuracy": max(
        experiment_history_2.history["val_accuracy"]
    ),
    "Best Validation Loss": min(
        experiment_history_2.history["val_loss"]
    ),
    "Test Accuracy": test_accuracy_2,
    "Test Loss": test_loss_2
})

results_df = pd.DataFrame(experiment_results)

results_df

Epoch 1/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.6261 - loss: 0.6616 - val_accuracy: 0.6769 - val_loss: 0.6042 - learning_rate: 5.0000e-04
Epoch 2/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6785 - loss: 0.6061 - val_accuracy: 0.7172 - val_loss: 0.5692 - learning_rate: 5.0000e-04
Epoch 3/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7089 - loss: 0.5760 - val_accuracy: 0.7245 - val_loss: 0.5585 - learning_rate: 5.0000e-04
Epoch 4/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7185 - loss: 0.5642 - val_accuracy: 0.7250 - val_loss: 0.5543 - learning_rate: 5.0000e-04
Epoch 5/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7260 - loss: 0.5571 - val_accuracy: 0.7286 - val_loss: 0.5521 - learning_rate: 5.0000e-04
Epoch 6/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7257 - loss: 0.5566 - val_accuracy: 0.7282 - val_loss: 0.5482 - learning_rate: 5.0000e-04
Epoch 7/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 

,Experiment,Learning Rate,Batch Size,Max Epochs,Best Validation Accuracy,Best Validation Loss,Test Accuracy,Test Loss
0,Experiment 1,0.0010,64,30,0.734381,0.543304,0.733619,0.547319
1,Experiment 2,0.0005,64,30,0.732857,0.543544,0.732762,0.548049


In [12]:
experiment_model_3 = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(64, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(32, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

experiment_model_3.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

experiment_history_3 = experiment_model_3.fit(
    X_train_scaled,
    y_train,
    validation_data=(X_val_scaled, y_val),
    epochs=30,
    batch_size=32,
    callbacks=[
        tf.keras.callbacks.EarlyStopping(
            monitor="val_loss",
            patience=8,
            restore_best_weights=True
        ),
        tf.keras.callbacks.ReduceLROnPlateau(
            monitor="val_loss",
            factor=0.5,
            patience=3,
            min_lr=1e-6
        )
    ],
    verbose=1
)

test_loss_3, test_accuracy_3 = experiment_model_3.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

experiment_results.append({
    "Experiment": "Experiment 3",
    "Learning Rate": 0.001,
    "Batch Size": 32,
    "Max Epochs": 30,
    "Best Validation Accuracy": max(
        experiment_history_3.history["val_accuracy"]
    ),
    "Best Validation Loss": min(
        experiment_history_3.history["val_loss"]
    ),
    "Test Accuracy": test_accuracy_3,
    "Test Loss": test_loss_3
})

results_df = pd.DataFrame(experiment_results)

results_df

Epoch 1/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.6653 - loss: 0.6206 - val_accuracy: 0.7210 - val_loss: 0.5643 - learning_rate: 0.0010
Epoch 2/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.7170 - loss: 0.5669 - val_accuracy: 0.7250 - val_loss: 0.5603 - learning_rate: 0.0010
Epoch 3/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7243 - loss: 0.5601 - val_accuracy: 0.7279 - val_loss: 0.5529 - learning_rate: 0.0010
Epoch 4/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7265 - loss: 0.5563 - val_accuracy: 0.7291 - val_loss: 0.5539 - learning_rate: 0.0010
Epoch 5/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7296 - loss: 0.5540 - val_accuracy: 0.7282 - val_loss: 0.5534 - learning_rate: 0.0010
Epoch 6/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7304 - loss: 0.5531 - val_accuracy: 0.7251 - val_loss: 0.5516 - learning_rate: 0.0010
Epoch 7/30
1532/1532 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7291 - l

,Experiment,Learning Rate,Batch Size,Max Epochs,Best Validation Accuracy,Best Validation Loss,Test Accuracy,Test Loss
0,Experiment 1,0.0010,64,30,0.734381,0.543304,0.733619,0.547319
1,Experiment 2,0.0005,64,30,0.732857,0.543544,0.732762,0.548049
2,Experiment 3,0.0010,32,30,0.734762,0.544566,0.731809,0.547585


In [13]:
experiment_model_4 = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(64, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(32, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.05),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

experiment_model_4.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

experiment_history_4 = experiment_model_4.fit(
    X_train_scaled,
    y_train,
    validation_data=(X_val_scaled, y_val),
    epochs=30,
    batch_size=64,
    callbacks=[
        tf.keras.callbacks.EarlyStopping(
            monitor="val_loss",
            patience=8,
            restore_best_weights=True
        ),
        tf.keras.callbacks.ReduceLROnPlateau(
            monitor="val_loss",
            factor=0.5,
            patience=3,
            min_lr=1e-6
        )
    ],
    verbose=1
)

test_loss_4, test_accuracy_4 = experiment_model_4.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

experiment_results.append({
    "Experiment": "Experiment 4",
    "Learning Rate": 0.001,
    "Batch Size": 64,
    "Max Epochs": 30,
    "Best Validation Accuracy": max(
        experiment_history_4.history["val_accuracy"]
    ),
    "Best Validation Loss": min(
        experiment_history_4.history["val_loss"]
    ),
    "Test Accuracy": test_accuracy_4,
    "Test Loss": test_loss_4
})

results_df = pd.DataFrame(experiment_results)

results_df

Epoch 1/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6662 - loss: 0.6179 - val_accuracy: 0.7076 - val_loss: 0.5764 - learning_rate: 0.0010
Epoch 2/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7193 - loss: 0.5642 - val_accuracy: 0.7267 - val_loss: 0.5564 - learning_rate: 0.0010
Epoch 3/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7268 - loss: 0.5548 - val_accuracy: 0.7271 - val_loss: 0.5510 - learning_rate: 0.0010
Epoch 4/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7299 - loss: 0.5508 - val_accuracy: 0.7259 - val_loss: 0.5501 - learning_rate: 0.0010
Epoch 5/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7298 - loss: 0.5491 - val_accuracy: 0.7278 - val_loss: 0.5505 - learning_rate: 0.0010
Epoch 6/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7324 - loss: 0.5474 - val_accuracy: 0.7284 - val_loss: 0.5491 - learning_rate: 0.0010
Epoch 7/30
766/766 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7312 - loss: 0.5466 - 

,Experiment,Learning Rate,Batch Size,Max Epochs,Best Validation Accuracy,Best Validation Loss,Test Accuracy,Test Loss
0,Experiment 1,0.0010,64,30,0.734381,0.543304,0.733619,0.547319
1,Experiment 2,0.0005,64,30,0.732857,0.543544,0.732762,0.548049
2,Experiment 3,0.0010,32,30,0.734762,0.544566,0.731809,0.547585
3,Experiment 4,0.0010,64,30,0.734286,0.542915,0.730857,0.547618


In [14]:
improved_model = experiment_model_3
improved_history = experiment_history_3

best_experiment = results_df.loc[
    results_df["Best Validation Accuracy"].idxmax()
]

print("Best Experiment:", best_experiment["Experiment"])
print("Best Validation Accuracy:", round(best_experiment["Best Validation Accuracy"] * 100, 2), "%")
print("Best Validation Loss:", round(best_experiment["Best Validation Loss"], 6))

Best Experiment: Experiment 3
Best Validation Accuracy: 73.48 %
Best Validation Loss: 0.544566


<div style="
    margin:22px 0 28px 0;
    padding:28px 32px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    letter-spacing:2px;
    margin-bottom:8px;
">
MODEL TRAINING & HYPERPARAMETER TUNING
</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:16px;
">
What We Did in Step 2
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.8;
    font-weight:600;
    margin:0 0 18px 0;
">
We trained the improved Dense Neural Network on the CardioML tabular
dataset and conducted a controlled set of hyperparameter experiments.
The goal was to determine which configuration provides the strongest
validation performance while keeping the model architecture appropriate
for the project data.
</p>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:17px 19px;
    margin-bottom:12px;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">
<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    margin-bottom:7px;
">
01 · EXPERIMENT DESIGN
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.7;
">
We kept the core Dense Neural Network architecture fixed and changed
selected hyperparameters such as learning rate and batch size. Each
experiment was trained for up to 30 epochs using Early Stopping and
ReduceLROnPlateau.
</div>
</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:17px 19px;
    margin-bottom:12px;
    border-radius:14px;
    background:#F8F5EE;
    border:1px solid #E5DCCB;
">
<div style="
    color:#8A6A3F;
    font-size:13px;
    font-weight:900;
    margin-bottom:7px;
">
02 · CONTROLLED EXPERIMENTS
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.7;
">
Four configurations were tested. Experiment 1 used a learning rate of
0.001 and batch size 64. Experiment 2 reduced the learning rate to
0.0005. Experiment 3 kept the learning rate at 0.001 and changed the
batch size to 32. Experiment 4 kept the learning rate and batch size
from Experiment 1 while reducing the dropout rates.
</div>
</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:17px 19px;
    margin-bottom:12px;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">
<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    margin-bottom:7px;
">
03 · MODEL SELECTION
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.7;
">
The experiments were compared using validation performance. Experiment 3
achieved the highest validation accuracy at 73.48%, so it was selected
as the best Sprint 2 configuration. The test set was kept separate from
the model-selection decision and used for final evaluation.
</div>
</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:18px 20px;
    border-radius:14px;
    background:#164E43;
">
<div style="
    color:#B7E4D2;
    font-size:13px;
    font-weight:900;
    letter-spacing:1px;
    margin-bottom:7px;
">
BEST CONFIGURATION
</div>

<div style="
    color:#FFFFFF;
    font-size:15px;
    font-weight:800;
    line-height:1.8;
">
Experiment 3 · Learning Rate: 0.001 · Batch Size: 32 · Max Epochs: 30
<br>
Best Validation Accuracy: 73.48%
<br>
Best Validation Loss: 0.544566
</div>
</div>

</div>

<div style="
    margin:22px 0 24px 0;
    padding:26px 30px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    display:flex;
    justify-content:space-between;
    align-items:center;
    gap:15px;
    margin-bottom:14px;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:2px;
">
STEP 3 · METRIC COMPARISON
</div>

<div style="
    padding:7px 13px;
    border-radius:10px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    color:#0F766E;
    font-size:12px;
    font-weight:900;
    white-space:nowrap;
">
SPRINT 2 REVIEW
</div>

</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:12px;
">
Comparing Model Performance
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.85;
    font-weight:600;
    margin:0;
    max-width:100%;
    box-sizing:border-box;
">
To evaluate the progress achieved during Sprint 2, the improved CardioML
model is compared with the Week 6 baseline and the Sprint 1 network.
All models are evaluated on the same CardioML classification task using
test accuracy as the primary comparison metric.
</p>

<div style="
    margin-top:18px;
    padding:17px 20px;
    border-radius:14px;
    background:#ECFDF5;
    border:1px solid #B7E4D2;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    letter-spacing:1px;
    margin-bottom:7px;
">
COMPARISON CRITERIA
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.7;
    font-weight:700;
    margin:0;
">
✓ Same CardioML task<br>
✓ Same evaluation approach<br>
✓ Week 6 baseline as the reference point<br>
✓ Sprint 2 represented by the best validated experiment
</p>

</div>

</div>

In [6]:
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

df = pd.read_csv("Inputs/heart_disease_uci.csv", sep=";")

X = df.drop("cardio", axis=1)
y = df["cardio"]

X_train, X_temp, y_train, y_temp = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=42,
    stratify=y
)

X_val, X_test, y_val, y_test = train_test_split(
    X_temp,
    y_temp,
    test_size=0.50,
    random_state=42,
    stratify=y_temp
)

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)
X_test_scaled = scaler.transform(X_test)

input_dim = X_train_scaled.shape[1]

improved_model = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(input_dim,)),
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(64, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.20),
    tf.keras.layers.Dense(32, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.10),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

improved_model.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

print("CardioML data and improved model are ready.")
print("Training shape:", X_train_scaled.shape)
print("Validation shape:", X_val_scaled.shape)
print("Test shape:", X_test_scaled.shape)

CardioML data and improved model are ready.
Training shape: (49000, 12)
Validation shape: (10500, 12)
Test shape: (10500, 12)


In [15]:
improved_model = experiment_model_3
improved_history = experiment_history_3

test_loss, test_accuracy = improved_model.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

print(f"Best Validation Accuracy: {max(improved_history.history['val_accuracy']):.4f}")
print(f"Best Validation Loss: {min(improved_history.history['val_loss']):.4f}")
print(f"Final Test Accuracy: {test_accuracy:.4f}")
print(f"Final Test Loss: {test_loss:.4f}")

Best Validation Accuracy: 0.7348
Best Validation Loss: 0.5446
Final Test Accuracy: 0.7318
Final Test Loss: 0.5476


In [16]:
import numpy as np
import pandas as pd

week6_accuracy = 0.7302
sprint1_accuracy = np.nan

sprint2_test_loss, sprint2_accuracy = improved_model.evaluate(
    X_test_scaled,
    y_test,
    verbose=0
)

metrics_comparison = pd.DataFrame({
    "Stage": [
        "Week 6 Baseline",
        "Sprint 1 Model",
        "Sprint 2 Model"
    ],
    "Task": [
        "CardioML - Tabular Classification",
        "CardioML - Tabular Classification",
        "CardioML - Tabular Classification"
    ],
    "Architecture": [
        "Dense Neural Network",
        "Sprint 1 Architecture",
        "Improved Dense Neural Network"
    ],
    "Test Accuracy": [
        week6_accuracy,
        sprint1_accuracy,
        sprint2_accuracy
    ]
})

metrics_comparison["Test Accuracy (%)"] = (
    metrics_comparison["Test Accuracy"] * 100
).round(2)

metrics_comparison

,Stage,Task,Architecture,Test Accuracy,Test Accuracy (%)
0,Week 6 Baseline,CardioML - Tabular Classification,Dense Neural Network,0.730200,73.02
1,Sprint 1 Model,CardioML - Tabular Classification,Sprint 1 Architecture,NaN,NaN
2,Sprint 2 Model,CardioML - Tabular Classification,Improved Dense Neural Network,0.731809,73.18


<div style="
    margin:22px 0 30px 0;
    padding:28px 32px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    letter-spacing:2px;
    margin-bottom:8px;
">
STEP 4 · SPRINT 2 COMPLETION
</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:14px;
">
Confirming Sprint 2 Completion
</div>

<p style="
    color:#304F47;
    font-size:15px;
    line-height:1.8;
    font-weight:600;
    margin:0 0 20px 0;
">
Before closing Sprint 2, the completed work must be committed to the
correct branch, reviewed through a Pull Request, and documented clearly
in the notebook. These checks confirm that the Sprint 2 deliverables are
ready for mentor review and final approval.
</p>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:16px 18px;
    margin-bottom:12px;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    margin-bottom:6px;
">
01 · WORK COMMITTED
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.6;
">
Commit all Sprint 2 notebook changes to the correct Git branch after
completing the experiments and documentation.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:16px 18px;
    margin-bottom:12px;
    border-radius:14px;
    background:#F8F5EE;
    border:1px solid #E5DCCB;
">

<div style="
    color:#8A6A3F;
    font-size:13px;
    font-weight:900;
    margin-bottom:6px;
">
02 · PR REVIEW
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.6;
">
Open a Pull Request and obtain mentor review and approval before merging
the Sprint 2 changes.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:16px 18px;
    margin-bottom:0;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    margin-bottom:6px;
">
03 · NOTEBOOK DOCUMENTED
</div>

<div style="
    color:#49665E;
    font-size:13px;
    font-weight:600;
    line-height:1.6;
">
Ensure that the architecture decision, experiments, results, metric
comparison, and Sprint 2 conclusions are clearly documented in the notebook.
</div>

</div>

</div>

<div style="
    margin:22px 0 30px 0;
    padding:28px 32px;
    border-radius:20px;
    background:linear-gradient(135deg,#F4FBF8 0%,#FFFFFF 55%,#F8F5EE 100%);
    border:1.5px solid #CFE5DC;
    box-shadow:0 8px 24px rgba(15,118,110,0.08);
    font-family:'Segoe UI',Arial,sans-serif;
    box-sizing:border-box;
    width:100%;
    max-width:100%;
    overflow:hidden;
">

<div style="
    color:#0F766E;
    font-size:13px;
    font-weight:900;
    letter-spacing:2px;
    margin-bottom:8px;
">
STEP 5 · SPRINT REVIEW & RETROSPECTIVE
</div>

<div style="
    color:#164E43;
    font-size:25px;
    font-weight:900;
    line-height:1.3;
    margin-bottom:18px;
">
Sprint Review & Retrospective
</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:18px 20px;
    margin-bottom:14px;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    margin-bottom:8px;
">
SPRINT REVIEW SUMMARY
</div>

<div style="
    color:#304F47;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
">
During Sprint 2, the CardioML core model was advanced using an improved
Dense Neural Network architecture and a controlled hyperparameter
experimentation process. Four configurations were trained and compared
using validation performance, with Experiment 3 selected as the best
configuration based on its highest validation accuracy of 73.48%.

The selected Sprint 2 model achieved a final test accuracy of 73.18%,
compared with 73.02% for the Week 6 baseline. This represents a
0.16 percentage-point improvement over the baseline. The experiments,
model selection process, and evaluation results were documented to
provide a clear basis for continued improvement.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:18px 20px;
    margin-bottom:14px;
    border-radius:14px;
    background:#FFFFFF;
    border:1px solid #D8E7E1;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    margin-bottom:8px;
">
WHAT WENT WELL
</div>

<div style="
    color:#304F47;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
">
• The architecture was selected according to the tabular nature of the CardioML dataset.<br>
• The core Dense Neural Network was improved with Batch Normalization and Dropout layers.<br>
• Four controlled hyperparameter configurations were trained and evaluated.<br>
• Validation performance was used to select the strongest configuration.<br>
• Training, validation, and test metrics were documented for comparison.<br>
• The selected Sprint 2 model achieved a measurable improvement over the Week 6 baseline.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:18px 20px;
    margin-bottom:14px;
    border-radius:14px;
    background:#F8F5EE;
    border:1px solid #E5DCCB;
">

<div style="
    color:#8A6A3F;
    font-size:14px;
    font-weight:900;
    margin-bottom:8px;
">
WHAT TO IMPROVE
</div>

<div style="
    color:#5F5140;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
">
• Explore a wider and more systematic hyperparameter search.<br>
• Investigate additional architectural configurations and regularization strategies.<br>
• Analyze whether further feature engineering can provide additional predictive information.<br>
• Monitor validation performance carefully to avoid overfitting during further tuning.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    padding:18px 20px;
    margin-bottom:0;
    border-radius:14px;
    background:#F4FBF8;
    border:1px solid #CFE5DC;
">

<div style="
    color:#0F766E;
    font-size:14px;
    font-weight:900;
    margin-bottom:8px;
">
ONE CONCRETE CHANGE FOR SPRINT 3
</div>

<div style="
    color:#304F47;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
">
In Sprint 3, I will implement a more systematic hyperparameter
optimization strategy that evaluates learning rate, batch size,
network depth, dropout, and training duration using validation
performance to identify the strongest configuration before performing
the final evaluation on the unseen test set.
</div>

</div>

</div>